lib/SIL/SILFunctionType.cpp

```diff
@@ -2143,10 +2143,17 @@ namespace {
 
     // SIL type lowering only does special things to tuples and functions.
 
-    /// Functions need to preserve their abstraction structure.
-    CanSILFunctionType visitSILFunctionType(CanSILFunctionType origType,
-                                            bool dropGenerics = false)
-    {
+    // When a function appears inside of another type, we only perform
+    // substitutions if it does not have a generic signature.
+    CanSILFunctionType visitSILFunctionType(CanSILFunctionType origType) {
+      if (origType->getGenericSignature())
+        return origType;
```

Comment:
But what if you have `(lambda (x) (x + y))` (where y comes from the outer generic context) and you substitute `y:=1`? I expect you should get `(lambda (x) (x+1))`.